backend/src/main/java/ai/verta/modeldb/experimentRun/ExperimentRunDAORdbImpl.java

```diff
@@ -2251,54 +2261,151 @@ public void deleteLogVersionedInputs(Session session, Long repoId, String commit
+    if (experimentRunObj == null) {
+      LOGGER.warn(ModelDBMessages.EXP_RUN_NOT_FOUND_ERROR_MSG);
+      Status status =
+          Status.newBuilder()
+              .setCode(Code.NOT_FOUND_VALUE)
+              .setMessage(ModelDBMessages.EXP_RUN_NOT_FOUND_ERROR_MSG)
+              .build();
+      throw StatusProto.toStatusRuntimeException(status);
+    }
+    Map<String, List<ArtifactEntity>> artifactEntityMap = experimentRunObj.getArtifactEntityMap();
+
+    List<ArtifactEntity> result =
+        (artifactEntityMap != null && artifactEntityMap.containsKey(ModelDBConstants.ARTIFACTS))
+            ? artifactEntityMap.get(ModelDBConstants.ARTIFACTS)
+            : Collections.emptyList();
+    return result.stream()
+        .filter(artifactEntity -> artifactEntity.getKey().equals(key))
+        .findFirst();
+  }
+
   @Override
   public Entry<String, String> getExperimentRunArtifactsS3PathAndMultipartUploadID(
       String experimentRunId, String key, long partNumber, S3KeyFunction initializeMultipart)
       throws ModelDBException {
     try (Session session = ModelDBHibernateUtil.getSessionFactory().openSession()) {
-      ExperimentRunEntity experimentRunObj =
-          session.get(ExperimentRunEntity.class, experimentRunId);
-      if (experimentRunObj == null) {
-        LOGGER.warn(ModelDBMessages.EXP_RUN_NOT_FOUND_ERROR_MSG);
+      ArtifactEntity artifactEntity = getArtifactEntity(session, experimentRunId, key);
+      return getS3PathAndMultipartUploadId(
+          session, artifactEntity, partNumber != 0, key, initializeMultipart);
+    }
+  }
+
+  public ArtifactEntity getArtifactEntity(Session session, String experimentRunId, String key)
+      throws ModelDBException {
+    Optional<ArtifactEntity> artifactEntityOptional =
+        getExperimentRunArtifact(session, experimentRunId, key);
+    return artifactEntityOptional.orElseThrow(
+        () -> new ModelDBException("Can't find specified artifact", io.grpc.Status.Code.NOT_FOUND));
+  }
+
+  private SimpleEntry<String, String> getS3PathAndMultipartUploadId(
+      Session session,
+      ArtifactEntity artifactEntity,
+      boolean partNumberSpecified,
+      String key,
+      S3KeyFunction initializeMultipart) {
+    String uploadId;
+    if (partNumberSpecified) {
+      uploadId = artifactEntity.getUploadId();
+      try {
+        String message = null;
+        if (uploadId == null) {
+          if (initializeMultipart == null) {
+            message = "Multipart wasn't initialized";
+          } else {
+            uploadId = initializeMultipart.apply(artifactEntity.getPath()).orElse(null);
+          }
+        }
+        if (uploadId == null) {
+          message = "Can't initialize multipart upload";
+        }
+        if (message != null) {
+          LOGGER.info(message);
+          throw new ModelDBException(message, io.grpc.Status.Code.FAILED_PRECONDITION);
```

Comment:
When Client sends `part_number` but multipart is not supported:
Instead of raising an error, could we change this to return a normal presigned URL (not multipart)?
And make sure `multipart_upload_ok` is false in the response. Thanks!

(based on my testing, it looks like we could just remove this one line, but I'd like confirmation from a backend eng)

Reply:
fixed. There's 2 validation checks: one is used during GetUrlForArtifact, another in commitMultipartArtifact.

I have removed the one for GetUrlForArtifact

Reply:
Makes sense to me!

Reply:
@convoliution to re-check

Reply:
I verified that 90e8007c126a1ff3af1387ebfd7faadb3342737f does the expected behavior (returns valid presigned URL and `multipart_upload_ok = False` w/ NFS setup)